ygot/pathstrings.go

```diff
@@ -0,0 +1,259 @@
+)
+
+// PathToString takes a gNMI Path and provides its string representation. For example,
+// the path Path{Element: []string{"one", "two", "three"} is converted to the string
+// "/one/two/three" and returned. Both the pre-0.4.0 "element"-based paths, and the
+//  >0.4.0 paths based on "elem" are supported. In the case that post-0.4.0 paths are
+// specified, keys that are specified in the path are concatenated onto the name of
+// the path element using the format [name=value]. If the path specifies both pre-
+// and post-0.4.0 paths, the pre-0.4.0 version is returned.
+func PathToString(path *gnmipb.Path) string {
+	var buf bytes.Buffer
+	buf.WriteRune('/')
+	if path.Element != nil {
+		for i, e := range path.Element {
+			buf.WriteString(e)
+			if i != len(path.Element)-1 {
+				buf.WriteRune('/')
+			}
+		}
+		return buf.String()
+	}
+
+	for i, e := range path.Elem {
+		buf.WriteString(e.Name)
+		if len(e.Key) != 0 {
+			var keys []string
+			for k := range e.Key {
+				keys = append(keys, k)
+			}
+			sort.Strings(keys)
+
+			for _, k := range keys {
+				buf.WriteString(fmt.Sprintf("[%s=%s]", k, e.Key[k]))
+			}
+		}
+		if i != len(path.Elem)-1 {
+			buf.WriteRune('/')
+		}
+	}
+	return buf.String()
+}
+
+// PathType is used to indicate a gNMI path type.
+type PathType int64
+
+const (
+	// StructuredPath represents a Path using the structured 'PathElem' message in
+	// the 'elem' field of the gNMI Path message.
+	StructuredPath PathType = iota
+	// StringSlicePath represents a Path using the 'Element' repeated string field
+	// of the gNMI path message.
+	StringSlicePath
+)
+
+// StringToPath takes an input string representing a path in gNMI, and converts
+// it to a gNMI Path message, populated with the specified path encodings.
+func StringToPath(path string, pathTypes ...PathType) (*gnmipb.Path, util.Errors) {
+	if len(pathTypes) == 0 {
+		return nil, []error{errors.New("no path types specified")}
+	}
+
+	pmsg := &gnmipb.Path{}
+	var errs util.Errors
+	for _, p := range pathTypes {
+		switch p {
+		case StructuredPath:
+			gp, err := StringToStructuredPath(path)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error building structured path: %v", err))
+				continue
+			}
+			pmsg.Elem = gp.Elem
+		case StringSlicePath:
+			gp, err := StringToStringSlicePath(path)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error building string slice path: %v", err))
+				continue
+			}
+			pmsg.Element = gp.Element
+		}
+	}
+
+	if errs != nil {
+		return nil, errs
+	}
+
+	return pmsg, nil
+}
+
+// StringToStringSlicePath takes a string representing a path, and converts it into a
+// gnmi.Path. For example, if the Path "/a/b[c=d]/e" is input, it is converted
+// to a gnmi.Path{Element: []string{"a", "b[c=d]", "e"}} which is returned.
+func StringToStringSlicePath(path string) (*gnmipb.Path, error) {
+	parsedPath := &gnmipb.Path{}
+
+	var inKey, inEscape bool
+	var buf bytes.Buffer
+	for i, ch := range path {
```

Comment:
Suggestion to simplify this function:
pe := strings.Split(path, "/") to get elements
for _, e := range pe {
k, v := ExtractKeyVal(e)
...
}

To get the key-value, do something like:
https://goplay.googleplex.com/p/aiqBFCWjRo
This includes some more error checking, though I'm not sure what the expectations are for the incoming string.

Reply:
btw. line 40 len(kv) != 2 should be an error case, not return nil.

Reply:
We can't do `strings.Split` in this code, because "/" can exist in an escaped path element, so we have to do per-character parsing of the path. I don't see a way to be able to use split since it wouldn't be aware of the escaping?

Reply:
I'm not clear that the function that you propose is actually more robust. How does it handle `/is/i[name=one\=two]` which is one of the test cases -- since in this case, what it'll do is @ line 39 it will split at the first "=". I think that you always need to have the previous character as you're parsing it to determine what the right thing to append is.

This approach if doing character by character parsing and keeping state isn't ideal, but with all the escaping cases that one needs to handle, I think you end up with something akin to this.

Reply:
Yeah, I realized some of these issues after seeing a few more possible string examples. However, per offline conversation, I still recommend splitting the processing into multiple steps to simplify the state machine e.g. https://goplay.googleplex.com/p/t2uMm9uyun